xwiki-platform-core/xwiki-platform-repository/xwiki-platform-repository-server-api/src/main/java/org/xwiki/repository/internal/RepositoryManager.java

```diff
@@ -996,9 +1033,66 @@ private boolean updateExtensionVersionDependencies(XWikiDocument document, Exten
         return needSave;
     }
 
+    public Extension resolveExtensionVersion(XWikiDocument extensionDocument, String extensionVersion) throws ResolveException {
+        BaseObject extensionObject = extensionDocument.getXObject(XWikiRepositoryModel.EXTENSION_CLASSREFERENCE);
+        if(extensionObject == null){
+            return null;
+        }
+        String extensionId = getValue(extensionObject, XWikiRepositoryModel.PROP_EXTENSION_ID, (String) null);
+
+        BaseObject extensionProxyObject = extensionDocument.getXObject(XWikiRepositoryModel.EXTENSIONPROXY_CLASSREFERENCE);
+        if(extensionProxyObject == null){
+            return null;
+        }
+        String repositoryId = getValue(extensionProxyObject, XWikiRepositoryModel.PROP_PROXY_REPOSITORYID, (String) null);
+
+        if(extensionId == null || repositoryId == null){
+            return null;
+        }
+
+        ExtensionRepository repository = this.extensionRepositoryManager.getRepository(repositoryId);
```

Comment:
Not sure why you do all this, the id of the version is stored in the version object itself so you don't need to search for it in repositories again (which is very expensive).